axum/src/middleware/map_request.rs

```diff
@@ -0,0 +1,462 @@
+use crate::response::{IntoResponse, Response};
+use axum_core::extract::{FromRequest, FromRequestParts};
+use futures_util::future::BoxFuture;
+use http::Request;
+use std::{
+    any::type_name,
+    convert::Infallible,
+    fmt,
+    future::Future,
+    marker::PhantomData,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Create a middleware from an async function that transforms a request.
+///
+/// This differs from [`tower::util::MapRequest`] in that it allows you to easily run axum specific
+/// extractors.
+///
+/// # Example
+///
+/// ```
+/// use axum::{
+///     Router,
+///     routing::get,
+///     middleware::map_request,
+///     http::Request,
+/// };
+///
+/// async fn set_header<B>(mut request: Request<B>) -> Request<B> {
+///     request.headers_mut().insert("x-foo", "foo".parse().unwrap());
+///     request
+/// }
+///
+/// async fn handler<B>(request: Request<B>) {
+///     // `request` will have an `x-foo` header
+/// }
+///
+/// let app = Router::new()
+///     .route("/", get(handler))
+///     .layer(map_request(set_header));
+/// # let _: Router = app;
+/// ```
+///
+/// # Rejection the request
+///
+/// The function given to `map_request` is allowed to also return a `Result` which can be used to
+/// rejection the request and return a response immediately, without calling the remaining
+/// middleware.
+///
+/// Specifically the valid return types are:
+///
+/// - `Request<B>`
+/// - `Request<Request<B>, E> where E:  IntoResponse`
+///
+/// ```
+/// use axum::{
+///     Router,
+///     http::{Request, StatusCode},
+///     routing::get,
+///     middleware::map_request,
+/// };
+///
+/// async fn auth<B>(request: Request<B>) -> Result<Request<B>, StatusCode> {
+///     let auth_header = request.headers()
+///         .get(http::header::AUTHORIZATION)
+///         .and_then(|header| header.to_str().ok());
+///
+///     match auth_header {
+///         Some(auth_header) if token_is_valid(auth_header) => Ok(request),
+///         _ => Err(StatusCode::UNAUTHORIZED),
+///     }
+/// }
```

Comment:
We have a new "simplest way to write an auth middleware", which feels like one of our most frequently asked questions 🥳

Reply:
I kinda wanna nuke tower-http's auth middleware. They're so much more complicated than just doing this.